vividus-plugin-azure-storage-account/src/main/java/org/vividus/azure/storage/blob/BlobStorageSteps.java

```diff
@@ -231,6 +237,106 @@ public void findBlobs(StringComparisonRule rule, String blobNameToMatch, String
+     *                          <li><b>STORY</b> - the variable will be available within the whole story,
+     *                          <li><b>NEXT_BATCHES</b> - the variable will be available starting from next batch
+     *                          </ul>
+     * @param variableName      The variable name to store the list of found blob names.
+     */
+    @When("I filter blobs by:$filter in container `$containerName` of storage account `$storageAccountKey`"
+            + " and save result to $scopes variable `$variableName`")
+    public void findBlobsByFilter(BlobFilter filter, String containerName,
+                                      String storageAccountKey, Set<VariableScope> scopes, String variableName)
+    {
+        filter.validate();
+        BlobContainerClient blobContainerClient = createBlobContainerClient(containerName, storageAccountKey);
+        bddVariableContext.putVariable(scopes, variableName, getLimitedBlobNames(blobContainerClient, filter));
+    }
+
+    private static List<String> getLimitedBlobNames(BlobContainerClient blobContainerClient, BlobFilter filter)
+    {
+        return filter.getResultsLimit()
+                .map(limit -> {
+                    ListBlobsOptions options = new ListBlobsOptions();
+                    options.setMaxResultsPerPage(limit.intValue());
+                    return filter.getBlobNamePrefix()
+                            .map(options::setPrefix)
+                            .map(opts -> getLimitPerPage(filter, limit,
+                                    blobContainerClient.listBlobsByHierarchy("/", opts, null)))
+                            .orElseGet(
+                                () -> getLimitPerPage(filter, limit,
+                                        blobContainerClient.listBlobs(options, null)));
+                })
+                .orElseGet(
+                    () -> filter.getBlobNamePrefix()
+                        .map(blobContainerClient::listBlobsByHierarchy)
+                        .orElseGet(blobContainerClient::listBlobs)
+                        .stream()
+                        .map(BlobItem::getName)
+                        .filter(filterMatched(filter))
+                        .collect(Collectors.toList())
+                );
+    }
+
+    private static List<String> getLimitPerPage(BlobFilter filter, Long limit, PagedIterable<BlobItem> pages)
+    {
+        List<String> blobNames = new ArrayList<>();
+        for (PagedResponse<BlobItem> page : pages.iterableByPage())
+        {
+            if (blobNames.size() < limit)
+            {
+                LOGGER.info("Blob Items on page: {}", page.getContinuationToken());
+                page.getValue()
+                        .stream()
+                        .map(BlobItem::getName)
+                        .filter(filterMatched(filter))
+                        .forEach(blobNames::add);
```

Comment:
In your realization you will add the items into the list and then remove them, I propose to rewrite logic to not add them when the limit is reached

Reply:
we could use simple for to stop adding elements to blobNames once it's size reaches limit